Freeze on Engine.removeAllEntities()

When I call Engine.removeAllEntities() my application freeze and eventually get out of memory:

com.badlogic.gdx.utils.GdxRuntimeException: java.lang.OutOfMemoryError: Java heap spaceat com.badlogic.gdx.backends.headless.HeadlessApplication$1.run(HeadlessApplication.java:94)
Caused by: java.lang.OutOfMemoryError: Java heap space
    at java.lang.reflect.Array.newArray(Native Method)
    at java.lang.reflect.Array.newInstance(Array.java:70)
    at com.badlogic.gdx.utils.reflect.ArrayReflection.newInstance(ArrayReflection.java:25)
    at com.badlogic.gdx.utils.Array.resize(Array.java:338)
    at com.badlogic.gdx.utils.Array.add(Array.java:100)
    at com.badlogic.ashley.core.Engine.removeEntity(Engine.java:131)
    at com.badlogic.ashley.core.Engine.removeAllEntities(Engine.java:143)

I have less than 100 active entities at this moment.

--- a/ashley/src/com/badlogic/ashley/core/Engine.java
+++ b/ashley/src/com/badlogic/ashley/core/Engine.java
@@ -143,8 +143,18 @@ public class Engine {
 	 * Removes all entities registered with this Engine.
 	 */
 	public void removeAllEntities() {
-		while(entities.size > 0) {
-			removeEntity(entities.first());
+		if (updating || notifying) {
+			for(Entity entity: entities) {
+				entity.scheduledForRemoval = true;
+			}
+			EntityOperation operation = entityOperationPool.obtain();
+			operation.type = EntityOperation.Type.RemoveAll;
+			entityOperations.add(operation);
+		}
+		else {
+			while(entities.size > 0) {
+				removeEntity(entities.first());
+			}
 		}
 	}
 	
@@ -385,6 +395,11 @@ public class Engine {
 			switch(operation.type) {
 				case Add: addEntityInternal(operation.entity); break;
 				case Remove: removeEntityInternal(operation.entity); break;
+				case RemoveAll:
+					while(entities.size > 0) {
+						removeEntityInternal(entities.first());
+					}
+					break;
 			}
 			
 			entityOperationPool.free(operation);
@@ -497,6 +512,7 @@ public class Engine {
 		public enum Type {
 			Add,
 			Remove,
+			RemoveAll
 		}
 		
 		public Type type;

